### ISSUE TITLE
Help text on form elements

### ISSUE THREAD

**Issue by biljanap** [wet-boew/wet-boew#6613]
Are there any recommendations regarding the placement and the treatment of help text on form elements?
Our application requires us to occasionally use help text on both individual form elements and groups of elements (grouped in FIELDSET). Any suggestions and/or examples that would ensure that help text is read in Forms mode (Assistive technology)?

**Comment by pjackson28** [comment]
If you want the help text read off in forms mode then the main options are:
1. Using the label element to associate the help text to a form field (1 to 1 relationship, should only use one label element per field since some screen readers have problems with multiple label elements).
2. Using the fieldset legend to associate the help text to multiple form fields (1 to many relationship)

**Comment by biljanap** [comment]
@pjackson28 this would work with short help text. Unfortunately our help text sometimes includes paragraphs and lists.

**Comment by biljanap** [comment]
Forgot to add that ideally help text would open at the user request (by selecting help button or link).

**Comment by pjackson28** [comment]
What about providing a link to the longer help text? Usually links will remain in the tabbing order when in forms mode.

**Comment by biljanap** [comment]
@pjackson28 I was thinking about that and placing my help text on the same page because we don't want the user to navigate away from the page. It might work but it does not look as a very sophisticated solution and it's more like something that belongs to pre-WET4 age :o). 
Would using a link to open a lightbox work? Would it be acceptable and accessible to place this link inside the label (before closing tag)?
Something like this:
```<thml>
<label for="someTextbox"><span class="field-name">Label text for some textbox</span><strong class="required">(required)</strong><a href="#openHelpText" aria-controls="centred-popup" class="wb-lbx" role="button">Help</a></label>

**Comment by RobJohnston** [comment]
I've been wondering the same.  The table of contents on the forms page in the style guide does link to help text but that anchor is not actually on the page (http://wet-boew.github.io/wet-boew-styleguide/v4/design/forms-en.html#forms-help-text).  

One example that I have seen is from StatsCan (but it's just short descriptions):  http://open.canada.ca/en/forms/submit-your-app

**Comment by biljanap** [comment]
@RobJohnston I don't see that StatsCanada is using any mechanism (at least not an obvious one) that would make their short help description accessible by forms mode users but I might be wrong.

**Comment by RobJohnston** [comment]
Nope, just the only example that I've seen of a form with descriptions.

**Comment by LaurentGoderre** [comment]
I think placing the link for the help inside the label could be problenatic because there is already a behavior associated to clicking a label (focusing on the form element it's associated too) and users with mobility issues probably rely on this behavior. Adding a link in there could be disruptive

@dfarough  @mysticjeff what do you think would work best? Adding a link after the label and before the form element or after the form element, or is there another option altogether you would prefer?

**Comment by biljanap** [comment]
@LaurentGoderre This is my concern as well. I tested the code I included before posting my issue. When part of the label element help link does not place the focus on form element. It opens lightbox but I admit it does not mean it's appropriate and accessible way of dealing with the issue. Any suggestion is welcome and very appreciated.

**Comment by biljanap** [comment]
Until somebody tells me that this is not acceptable and offers better solution I am going with help link inside label. The link will be opening lightbox with help text. Thanks.

**Comment by dfarough** [comment]
I don't understand what you are trying to do.  You want to open or move focus to some helptext related to a particular field in a form?  Why is it necessary for the user to view this content in forms mode?

would this text be presented in a new window or tab?  Is this a better alternative?

**Comment by LaurentGoderre** [comment]
@dfarough the goal is to have a link to contextual help for certain fields. What is the best way to provide such links?

**Comment by dfarough** [comment]
you could use a graphic with appropriate alt text IE Help for field

**Comment by LaurentGoderre** [comment]
Yes, but where in the DOM should it be placed.

**Comment by biljanap** [comment]
....and also where to place help text?

**Comment by dfarough** [comment]
I agree that the label for a field should not be used to contain a link to open context sensitive help information for a field.  Instead, a link or button should be provided following the field for this purpose.  Depending on the amount and type of information being provided, perhaps the WET expandable collapsible content functionality could be used for this purpose.  I see no need to insure that the use of forms mode is maintained for users of access technology.  In fact using a button might be even better than a link since it would show up when listing all form fields in a page.

If you are providing many instances of this type of help, insure that the field context is specified first when defining the button so that users know immediately the context of the help being provided.

**Comment by RobJohnston** [comment]
I think that an example of such a WET expandable-collapsible help text, although currently for WET 3, can be found here:  [Case Specific Enquiry](https://dmp-portal.cic.gc.ca/enquiries-renseignements/canada-case-cas-eng.aspx).  

Full disclosure, I have this on my plate to update to WET 4.

Starting with the form basic example, would this be suitable markup?  Comments anyone?
```html
<form role="form" method="get" action="#">
    <div class="text-left">
        <div class="row">
            <div class="col-xs-12 col-md-4">
                <div class="form-group">
	                <label for="exampleInputEmail1">Email address</label>
	                <input type="email" class="form-control" id="exampleInputEmail1" placeholder="Enter email" />
                </div>
            </div>
            <div class="col-xs-6 col-md-8">
                <details>
                    <summary><span class="glyphicon glyphicon-question-sign"></span></summary>
                    <div class="panel panel-info">
                        <div class="panel-body">
                            <p>E-mail (electronic mail) is a system that allows you to send and receive electronic messages between computers.</p>
                        </div>
                    </div>
                </details>
            </div>
        </div>
        <div class="row">
            <div class="col-xs-12 col-md-4">
                <div class="form-group">
	                <label for="exampleInputPassword1">Password</label>
	                <input type="password" class="form-control" id="exampleInputPassword1" placeholder="Password" />
                </div>
            </div>
            <div class="col-xs-6 col-md-8">
                <details>
                    <summary><span class="glyphicon glyphicon-question-sign"></span></summary>
                    <div class="panel panel-info">
                        <div class="panel-body">
                            <p>A password is a sequence of characters that one must input to gain access to a file, application, or computer system.</p>
                        </div>
                    </div>
                </details>
            </div>
        </div>            
    </div>
    <div class="form-group">
	    <label for="exampleInputFile">File input</label>
	    <input type="file" id="exampleInputFile" />
    </div>
    <div class="checkbox">
	    <label><input type="checkbox" /> Check me out</label>
    </div>
    <button type="submit" class="btn btn-default">Submit</button>
</form>
```

**Comment by robwloch** [comment]
@RobJohnston , thank you for working on this. In our department, we've been struggling to figure out a nice way to have workable field help. We've run into strange issues with screen readers when we have a radio button or checkbox followed by a link, followed by an unrelated fieldset as it doesn't read the legend of that fieldset. Not sure if it's an NVDA bug or not.

I ran your code locally and it worked nicely for me, then I tried with NVDA and it also worked well but when tabbing to the question mark it just read "button collapsed" so no indication that it's field help so obviously we'd add some sr-only text like "help with email".

I will play with your code by adding a fieldset and radio buttons to see if those inputs work nicely too or suffer from the screen reader issue.

**Comment by dfarough** [comment]
with regard to choosing text to be used for the control - Specify the field or relationship first followed by the purpose of the button.  So to clarify, for this example, you might use E-Mail help" .  Keep text short and to the point.  If prompting is too verbose, people may be  tempted to tab away before the important information is spoken.  screen readers offer the ability to generate a list of numerous elements on the page.  a user could generate a list of buttons or fields present on the page and then move directly to the element in question by pressing the enter key.

**Comment by martynka** [comment]
Screen reader test didn’t demonstrate much improvement. The original form says: “Applicant Information Unlabeled 0 toggle button collapse” the improved form says “Unlabeled 1 button collapsed”. The text is not sufficient to know what the collapsible buttons mean.

 As far as the coding goes, other than the empty summary element I don’t see anything else that’s out of ordinary (others in the community should confirm).

**Comment by RobJohnston** [comment]
So from the comments above, and without trying it, I think the markup of the ```<summary>``` tag should be changed to something like the following:
```html
<summary><span class="glyphicon glyphicon-question-sign wb-inv">E-mail help</span></summary>
```
This uses the [Screen reader invisible content](http://wet-boew.github.io/wet-boew-styleguide/v4/design/visible-en.html#invisible).  Would that solve the problem?

**Comment by LaurentGoderre** [comment]
Are you sure this actually works? I never tried to combine a glyphicon with `wb-inv` before

**Comment by RobJohnston** [comment]
Oh, gosh.... Friday afternoon...that's not going to work.  The closest that I can guess is this:
```html
<summary>
  <span class="glyphicon glyphicon-question-sign"></span>
  <span class="wb-inv">E-mail help</span>
</summary>
```

**Comment by martynka** [comment]
The code proposed works, tested with JAWS and NVDA. Both screen readers speak: "email help button collapsed" and "password help button collapsed".

**Comment by RobJohnston** [comment]
That's great.  I created a gist with the original and the edit if anybody else wants to copy-paste and test some more.  The gist can be found here:  https://gist.github.com/RobJohnston/099c412aff770181cba4

When it's all nailed down, the documentation should be updated here:
http://wet-boew.github.io/wet-boew-styleguide/v4/design/forms-en.html#forms-help-text
There's already a link in the table of contents but no actual content.

**Comment by robwloch** [comment]
I haven't had time yet to try fieldsets with several radio buttons or checkboxes but that is something that would be good to add to the example. Also, it seems to me that the help icons are quite far from the labels which is good for neat alignment but not sure if any accessibility concerns for users with screen magnifiers for example?

**Comment by RobJohnston** [comment]
I've updated the gist with the radio button list from the [form validation example](http://wet-boew.github.io/v4.0-ci/demos/formvalid/formvalid-en.html) and help text to match.

**Comment by robwloch** [comment]
Thank you @RobJohnston . It works nicely in NVDA. We have some apps that have help links on individual radio buttons and checkboxes within fieldsets. Do you think that would be possible with this design or better to keep help only at the legend level? 

ie.

![capture](https://cloud.githubusercontent.com/assets/5721679/7254485/68164836-e811-11e4-9631-f77da6e853db.PNG)

**Comment by RobJohnston** [comment]
Interesting... I don't know (I'm usually a back-end developer).  That way solves the outstanding problem you mentioned of the help icons being far from the labels.

**Comment by robwloch** [comment]
Our organization is currently using the following lightbox solution for modal contextual help. I know that earlier in this thread the lightbox was suggested but didn't really gain any traction. Do you think it could be used as an alternate solution? Of course, the link should probably not be in the label.

<label for="postalCode" class="required">Postal code or ZIP code<strong class="required">(required)</strong><a href="#postalCodeHelp" title="Postal code or ZIP code" aria-controls="centred-popup" class="wb-lbx" role="button">
	<span class="glyphicon glyphicon-question-sign">
		<span class="wb-inv">#postalCodeHelp</span>
	</span>
</a>
</label>

![lightbox0](https://cloud.githubusercontent.com/assets/5721679/7255379/f54fec66-e816-11e4-961a-34b4d6d9868f.PNG)

![lightbox1](https://cloud.githubusercontent.com/assets/5721679/7255383/fa1d09ae-e816-11e4-89b0-3ad767009b80.PNG)

**Comment by hsrudnicki** [comment]
Are you ready to close this issue?

**Comment by hsrudnicki** [comment]
Made "Going Stale" to close this issue,  referenced with the others issues (3 opened and one closed) or if otherwise? Either way, please update this issue's status before closing if no further discussion continued, thank you

**Comment by RubyDo** [comment]
@RobJohnston I love the code. Thank you! The only incontinence is that screen reader user will get to the help text after the input field which might be late. Therefore, I added an association between the input field and the help text using aria-describedby so that the screen reader reads the instruction before  edit of the input field. 

<div class="row">
            <div class="col-xs-12 col-md-4">
                <div class="form-group">
                    <label for="exampleInputEmail1">Email address</label>
                    <input type="email" class="form-control" id="exampleInputEmail1" placeholder="Enter email" aria-describedby="helpemail" />
                </div>
            </div>
            <div class="col-xs-6 col-md-8">
                <details>
                    <summary><span class="glyphicon glyphicon-question-sign wb-inv">E-mail help</span></summary>
                    <div class="panel panel-info">
                        <div class="panel-body" id="helpemail">
                            <p>E-mail (electronic mail) is a system that allows you to send and receive electronic messages between computers.</p>
                        </div>
                    </div>
                </details>
            </div>
        </div>

**Comment by RobJohnston** [comment]
@RubyDo I think that's valuable... just the kind of thinking that this discussion needs.  I've updated the [gist](https://gist.github.com/RobJohnston/099c412aff770181cba4), but where should the aria-describedby be placed for the radio button list?

**Comment by RubyDo** [comment]
@RobJohnston aria-describedby would be placed in the <input> of each radio button. I forked the gist and updated it on https://gist.github.com/RubyDo/fdba23b5e77a5f998d59 (couldn't know how to submit this to you on GitHub - still newbie).

Here is a screen reader demo reading the instruction associated with each form control (including the radio buttons) https://youtu.be/ZpqmcjFJY4g

**Comment by hsrudnicki** [comment]
Is this still an issue? Either way, please update this issue's status, whether this issue can be closed as resolved/fixed or if otherwise, thank you

**Comment by RobJohnston** [comment]
Just want to note that currently WET includes the `help-block` class from BootStrap (http://getbootstrap.com/css/#forms-help-text). Using this, the help text receives the validation styles (so it turns red on error, for example).  

There may be opportunity to use this class in the options already mentioned.

However, in BootStrap 4, `help-text` is being replace with `text-muted`.

**Comment by RubyDo** [comment]
Thanks @RobJohnston for sharing the info

**Comment by hsrudnicki** [comment]
Is this still an issue? Can be this closed as resolved or otherwise?

**Comment by kodecount** [comment]
This was furthered discussed here: https://github.com/wet-boew/wet-boew/issues/7896

I would say this issue can be closed (instead of having multiple of the same issue).
